go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,374 @@
+
+```go
+type MyInterface interface {
+  DoSomething() error
+  DoAnotherThing() error
+  mustEmbedBase()
+}
+
+type MyInterfaceBase struct{}
+func (MyInterfaceBase) DoSomething() error { return nil }
+func (MyInterfaceBase) DoAnotherThing() error { return nil }
+func (MyInterfaceBase) mustEmbedBase() {}
+```
+
+The library knows all implementations of the interface embedded the struct they had to for forwards compatibility and it
+knows that nobody's compilation will fail because they didn't embed the base. This is what gRPC and other libraries are
+starting to do to force forward compatibility of interfaces.
+
+### WorkerInterceptor
+
+Current:
+
+```go
+type WorkflowInterceptor interface {
+  InterceptWorkflow(info *WorkflowInfo, next WorkflowInboundCallsInterceptor) WorkflowInboundCallsInterceptor
+}
+```
+
+Proposal:
+
+```go
+type WorkerInterceptor interface {
+  InterceptActivity(ctx context.Context, next ActivityInboundInterceptor) ActivityInboundInterceptor
+  InterceptWorkflow(ctx Context, next WorkflowInboundInterceptor) WorkflowInboundInterceptor
+}
+```
+
+Then change `WorkerOptions.WorkflowInterceptorChainFactories []WorkflowInterceptor` to
+`WorkerOptions.Interceptors []WorkerInterceptor`.
+
+For discussion:
+
+* This maps to what Java does
+* Would we want a `WorkerInterceptorBase` with default impls that return instances of `ActivityInboundInterceptorBase`
+  and `WorkflowInboundInterceptorBase`? Why doesn't Java have this?
+
+## Tracing and Context Propagation
+
+Current:
+
+* A concept called a context propagator handles marshalling to/from workflow and activity proto "headers" (a construct
+  in the body not to be confused with actual HTTP headers)
+* Client options accept an OpenTracing tracer
+* An internal context propagator is creating for the OpenTracing tracer that marshals to/from a span context to a
+  special key in the headers
+* The tracer is sent all throughout the contextual work of workflows and activities so that spans can be created at
+  different times
+
+Proposal:
+
+* A new ability to retrieve a mutable set of proto "headers" on the context will be made so that interceptors can set
+  workflow/activity headers in either direction
+* Internally, headers will be obtained from the context by the last interceptor
+  * So headers are set on the context before interceptor invocation for mutation by interceptors
+* A `tracing` package will be created with:
+  * `WorkerInterceptor` implementations for OpenTracing and OpenTelemetry tracers
+    * Wraps the same inbound/outbound calls in spans that are otherwise wrapped by the tracer today
+    * On inbound `ExecuteActivity` and `ExecuteWorkflow` span context is unmarshaled from header
+    * On outbound `ExecuteActivity`, `ExecuteLocalActivity`, `ExecuteChildWorkflow`, and `NewContinueAsNewError` span
+      context is marshaled into header
+  * `Client` implementations for OpenTracing and OpenTelemetry tracers
```

Comment:
The other SDKs have a concept of `WorkflowClientCallsInterceptor`s.
While the same functionality can be achieved with the current proposal it might be worth considering adding the interceptor for consistency with the other SDKs.

Reply:
I have added a section on why this is unnecessary. Open for comment.